### PR TITLE
Changes that can go in independently from the "parenthese preferences" PR

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
These were all found while working on https://github.com/dotnet/roslyn/pull/26013.  However, they are not strictly part of that PR.  To help keep that PR simpler, i've pulled them out into their own change.  https://github.com/dotnet/roslyn/pull/26013 is then strictly about adding the feature.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv and @sharwell .  Reviewing this PR independently will make reviewing https://github.com/dotnet/roslyn/pull/26013 much simpler.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jmarolf as well.

File: src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs [hunk 1]
```diff
@@ -591,7 +591,7 @@ protected static (OptionKey, object) SingleOption<T>(PerLanguageOption<CodeStyle
             var result = new Dictionary<OptionKey, object>();
             foreach (var option in options)
             {
-                result.Add(option.key, option.value);
+                result[option.key] = option.value;
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs]: ❗️ Previous code was correct

**Comment by CyrusNajmabadi** [comment]
> Needs to be refactored so each commit is a single conceptual operation which transforms the code base from valid to valid, and includes a description of the change.

Sure, i'll see what i can do.

**Comment by CyrusNajmabadi** [comment]
@sharwell have created separate commits for the changes.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -283,41 +322,24 @@ private static bool RemovalChangesAssociation(ParenthesizedExpressionSyntax node
...
+                    //
+                    // i.e. if you have "a && (b && c)" it can be converted to "a && b && c" 
+                    // as that new interpretation "(a && b) && c" operates the exact same way at 
+                    // runtime.
+                    //
+                    // Specifically: 
+                    //  1) the operands are still executed in the same order: a, b, then c.
+                    //     So even if they have side effects, it will not matter.
+                    //  2) the same shortcircuiting happens.
+                    //  3) the result will always be the same (for logical operators, there are
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: :bulb: Move 'for logical operators' before the opening parenthesis for improved readability

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 3]
```diff
@@ -249,7 +286,9 @@ private static bool RemovalMayIntroduceInterpolationAmbiguity(ParenthesizedExpre
             return false;
         }
 
-        private static bool RemovalChangesAssociation(ParenthesizedExpressionSyntax node, ExpressionSyntax expression, ExpressionSyntax parentExpression, SemanticModel semanticModel)
+        private static bool RemovalChangesAssociation(
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: :memo: The unnecessary wrapping of an unchanged line caused this review to take longer than it otherwise needed to.

File: src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            // of the parentheses might break semantics. Note that we'll try and be 
+            // fairly conservative with these. For example, we'll assume that failing 
+            // any of these checks results in the parentheses being declared as necessary 
+            // -- even if they could be removed depending on whether the parenthesized
+            // expression appears on the left or right side of the parent binary expression.
+
+            // First, does the binary expression result in an operator overload being 
+            // called?
+            var symbolInfo = semanticModel.GetSymbolInfo(innerExpression);
+            if (AnySymbolIsUserDefinedOperator(symbolInfo))
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: :memo: (for other reviewers) This is similar to the original, but has been updated to return "not safe to change associativity" for cases where multiple candidate symbols exist and at least one of the candidates meets the condition.

File: src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            var innerTypeInfo = semanticModel.GetTypeInfo(innerExpression);
+            if (innerTypeInfo.Type != null && innerTypeInfo.ConvertedType != null)
+            {
+                if (!innerTypeInfo.Type.Equals(innerTypeInfo.ConvertedType))
+                {
+                    return false;
+                }
+            }
+
+            // Floating point is not safe to change associativity of.  For example,
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: :memo: (for other reviewers) This section is new

❓ Does this need to special-case the handling of `int + (int + long)`?

File: src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Extensions
+{
+    internal static class CommonParenthesizedExpressionSyntaxExtensions
+    {
+        public static bool IsSafeToChangeAssociativity(
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: :memo: (for other reviewers) This method is extracted from code in `RemovalChangesAssociation`, but has been updated to invert the meaning of the return value and handle some additional cases noted in the diff.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 7]
```diff
@@ -2621,6 +2623,9 @@ public static OperatorPrecedence GetOperatorPrecedence(this ExpressionSyntax exp
                 case SyntaxKind.CheckedExpression:
                 case SyntaxKind.UncheckedExpression:
                 case SyntaxKind.AnonymousMethodExpression:
+                // unsafe code
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: ❓ Is this part of an unrelated change?

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs [hunk 8]
```diff
@@ -41,10 +39,46 @@ public static bool IsRegularOrDocumentationComment(this ISyntaxFactsService synt
...
+            return resultNode;
+        }
+
         public static bool SpansPreprocessorDirective(this ISyntaxFactsService service, SyntaxNode node)
             => service.SpansPreprocessorDirective(SpecializedCollections.SingletonEnumerable(node));
 
         public static bool IsWhitespaceOrEndOfLineTrivia(this ISyntaxFactsService syntaxFacts, SyntaxTrivia trivia)
             => syntaxFacts.IsWhitespaceTrivia(trivia) || syntaxFacts.IsEndOfLineTrivia(trivia);
+
+        public static void GetPartsOfBinaryExpression(
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs]: ❓ I don't see where `GetPartsOfBinaryExpression` and `GetExpressionOfParenthesizedExpression` are used here. Were they supposed to be in a different commit?

:bulb: This method should be named `GetExpressionsOfBinaryExpression` for consistency with the next method. It might further make sense to return `(SyntaxNode left, SyntaxNode right)` instead of using `out` variables.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 9]
```diff
@@ -45,25 +45,27 @@ public static ExpressionSyntax WalkDownParentheses(this ExpressionSyntax express
             return expression;
         }
 
-        public static ExpressionSyntax Parenthesize(this ExpressionSyntax expression, bool includeElasticTrivia = true)
+        public static ExpressionSyntax Parenthesize(
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: :memo: This was an unnecessarily complex way to add the `addSimplifierAnnotation` parameter, but now that I've reviewed it, it might as well stay. Do keep in mind that every time something like this happens, things just take more time.

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 10]
```diff
@@ -818,9 +816,16 @@ public SyntaxToken ToIdentifierToken(string name)
             return name.ToIdentifierToken();
         }
 
-        public SyntaxNode Parenthesize(SyntaxNode expression, bool includeElasticTrivia)
+        public SyntaxNode Parenthesize(SyntaxNode expression, bool includeElasticTrivia, bool addSimplifierAnnotation)
+            => ((ExpressionSyntax)expression).Parenthesize(includeElasticTrivia, addSimplifierAnnotation);
+
+        public void GetPartsOfParenthesizedExpression(
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: :memo: There seems to be some mixed code between this and the previous commit.

File: src/Features/Core/Portable/CodeStyle/AbstractCodeStyleDiagnosticAnalyzer.cs [hunk 11]
```diff
@@ -60,19 +60,25 @@ internal abstract class AbstractCodeStyleDiagnosticAnalyzer : DiagnosticAnalyzer
             WarningDescriptor = CreateDescriptorWithSeverity(DiagnosticSeverity.Warning);
             ErrorDescriptor = CreateDescriptorWithSeverity(DiagnosticSeverity.Error);
 
-            UnnecessaryWithSuggestionDescriptor = CreateDescriptorWithId(
-                descriptorId, _localizableTitle, _localizableMessage, 
-                DiagnosticSeverity.Hidden, DiagnosticCustomTags.Unnecessary);
+            UnnecessaryWithSuggestionDescriptor = CreateUnnecessaryDescriptor(DiagnosticSeverity.Hidden);
 
             UnnecessaryWithoutSuggestionDescriptor = CreateDescriptorWithId(
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/CodeStyle/AbstractCodeStyleDiagnosticAnalyzer.cs]: Done!

File: src/Workspaces/CSharp/Portable/CodeGeneration/ExpressionGenerator.cs [hunk 12]
```diff
@@ -79,17 +79,17 @@ private static ExpressionSyntax GenerateNullLiteral()
...
-                case byte val: return GenerateLiteralExpression(type, val, LiteralSpecialValues.ByteSpecialValues, null, canUseFieldReference, (s, v) => SyntaxFactory.Literal(s, v));
-                case ushort val: return GenerateLiteralExpression(type, val, LiteralSpecialValues.UInt16SpecialValues, null, canUseFieldReference, (s, v) => SyntaxFactory.Literal(s, (uint)v));
-                case uint val: return GenerateLiteralExpression(type, val, LiteralSpecialValues.UInt32SpecialValues, null, canUseFieldReference, SyntaxFactory.Literal);
-                case ulong val: return GenerateLiteralExpression(type, val, LiteralSpecialValues.UInt64SpecialValues, null, canUseFieldReference, SyntaxFactory.Literal);
+                case sbyte val: return GenerateNonNegativeLiteralExpression(type, val, LiteralSpecialValues.SByteSpecialValues, null, canUseFieldReference, (s, v) => SyntaxFactory.Literal(s, v));
+                case short val: return GenerateNonNegativeLiteralExpression(type, val, LiteralSpecialValues.Int16SpecialValues, null, canUseFieldReference, (s, v) => SyntaxFactory.Literal(s, v));
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/ExpressionGenerator.cs]: because i'm dumb.  #fixed.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 13]
```diff
@@ -29,6 +38,17 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node,
                 return true;
             }
 
+            if (node.IsParentKind(SyntaxKind.ArrowExpressionClause))
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: Done.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 14]
```diff
@@ -29,6 +38,17 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node,
...
+            {
+                return true;
+            }
+
+            // checked((x)) -> checked(x)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: yup.  those cases are all handled below in hte section with this comment:

```c#
            // Easy statement-level cases:
            //   var y = (x);           -> var y = x;
            //   var (y, z) = (x);      -> var (y, z) = x;
            //   if ((x))               -> if (x)
            //   return (x);            -> return x;
            //   yield return (x);      -> yield return x;
            //   throw (x);             -> throw x;
            //   switch ((x))           -> switch (x)
            //   while ((x))            -> while (x)
            //   do { } while ((x))     -> do { } while (x)
            //   for(;(x);)             -> for(;x;)
            //   foreach (var y in (x)) -> foreach (var y in x)
            //   lock ((x))             -> lock (x)
            //   using ((x))            -> using (x)
            //   catch when ((x))       -> catch when (x)
```

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: done.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: My editor window is most of my screen and already shows 160 chars without problem.  When it goes beyond that, it actually makes the code harder to read with extra scrolling.  I don't have a problem with a soft limit of around that amount so that code can be viewed on one screen without that being arduous on the authors.  i.e. i'm not asking for 80 columns or something so short that people are always running into it :)

File: src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: Probably.  That may be caught by the above check.  But it may not be.  I'm going to add a check that all the types have to be the same.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: #done.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Sorry, could you clarify?  I'm not seeing a simpler way to do things?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: No.  These were part of hte change where i check if it's ok to remove parens and i need to get operator precedence to do so.  We were missing operator precedence for unsafe operators (like ```*x```), which made us not know properly if we could, for example, remove parens in ```(void*)(*x)```

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs]: The naming consistency is either 'Parts' to indicate the deconstruction when there are multiple things returned, or 'GetXXXOfYYY' where 'XXX' is the name of the actual child property if only one child is being returned.  This is consistent with the rest of hte naming in this file.

**Comment by CyrusNajmabadi** [comment]
@sharwell changes have been made.

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs [hunk 15]
```diff
@@ -61,13 +61,13 @@ public void TestLiteralExpressions()
         {
             VerifySyntax<LiteralExpressionSyntax>(Generator.LiteralExpression(0), "0");
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: tests added.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: Yep, I'm not saying you have to change it, but I've decided to point out aspects of the PR which caused the review process to be slower. Similar situations can be avoided in the future to improve PR throughput.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs]: Note: now that we have tuples, these "GetPartsOfXXX" members will likely become DeconstructXXX methods in teh future.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 3 above)
Comment (Neme12) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: On the other hand, it's going to take less time when anyone makes changes here in the future, such as adding a parameter.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: I think this comes down to a personal preference take.  I find your code to be less clear.  But i def can appreciate that different patterns and styles can impact the review.  I'll try to find an appropriate balance that allows for expressiveness and cleanliness, without overdoing it on the PR side.

**Comment by CyrusNajmabadi** [comment]
> I'm sorry but is it that bad if it takes just a few more seconds? Throughput is slow because of time it takes to even get some eyes on a PR, not because of extra 10 seconds when looking at the code.

I think Sam is asking for this to be part of hte consideration when making the PR.  For example, i could have done the change in two commits.  One with the logic change, one with the cleanup change.  That could have made things easier and smoother to review.

he's right in that PRs have a cost to them, and a good balance needs to be struck around both the value of hte change and the cost to get it in.  I try to have a reasonable balance there in practice.   But there is always room for improvement!

**Comment by CyrusNajmabadi** [comment]
> On the other hand, it's going to take less time when anyone makes changes here in the future, such as adding a parameter.

This is personally one of those areas where i wouldn't mind the team just setting on some style with some automated system to keep things in line.  As mentioned previously, i think it's valuable when code can be viewed on github without excessive horizontal scrolling.  If we pick a reasonable resolutoin that our data tells us is the most common for devs, and we check that against github, i think we can come up with a reasonable boundary that both ensures code is always visible, without making people feel excessively cramped.  In my own investigation on this, it looks like that's around 130-140 chars.

Fortunately, this is also an amount that seems somewhat reasonable for C#.  i.e. something like 80, or even 100, often feels like you're butting up against it *all the time*.  However, past 130, it's quite rare, and is just these long parameter lists.  

This feels like a ripe area for just some automated styling, which would then make this a non-issue for most PRs going forward.  You'd only really hit it if you were making the change that took things past the limit.

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs [hunk 16]
```diff
@@ -121,6 +121,44 @@ public void TestLiteralExpressions()
...
+            VerifySyntax<MemberAccessExpressionSyntax>(Generator.LiteralExpression(sbyte.MinValue), "global::System.SByte.MinValue");
+            VerifySyntax<MemberAccessExpressionSyntax>(Generator.LiteralExpression(sbyte.MaxValue), "global::System.SByte.MaxValue");
+        }
+
+        [Fact]
+        public void TestByteLiteralExpressions()
+        {
+            VerifySyntax<LiteralExpressionSyntax>(Generator.LiteralExpression((byte)0), "0");
+            VerifySyntax<LiteralExpressionSyntax>(Generator.LiteralExpression((byte)1), "1");
+            VerifySyntax<LiteralExpressionSyntax>(Generator.LiteralExpression(byte.MinValue), "0");
```
Comment (sharwell) [inline src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: ❓ Why is this not `global::System.Byte.MinValue`?

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: From LiteralSpecialValues.cs:

```c#
        // Let's not have special values for byte. byte.MaxValue seems overkill versus 255.
        public static readonly IEnumerable<KeyValuePair<byte, string>> ByteSpecialValues = new Dictionary<byte, string>();
```

I did not think it worthwhile to change that logic :)

**Comment by jinujoseph** [comment]
@CyrusNajmabadi  could you resolve the conflict + @sharwell  may trigger another review of test ?

**Comment by CyrusNajmabadi** [comment]
Will do.

**Comment by CyrusNajmabadi** [comment]
Conflicts resolved.

**Comment by CyrusNajmabadi** [comment]
@jinujoseph @sharwell @jcouv please merge when convenient.

**Comment by CyrusNajmabadi** [comment]
Thanks!
